master/internal/db/postgres_experiments.go

```diff
@@ -35,3 +38,148 @@ func (db *PgDB) ExperimentLabelUsage() (labelUsage map[string]int, err error) {
 	}
 	return labelUsage, nil
 }
+
+// MetricNames returns the set of training and validation metric names that have been recorded for
+// an experiment.
+func (db *PgDB) MetricNames(experimentID int, sStartTime time.Time, vStartTime time.Time) (
+	training []string, validation []string, sEndTime time.Time, vEndTime time.Time, err error) {
+	type namesWrapper struct {
+		Name    string    `db:"name"`
+		EndTime time.Time `db:"end_time"`
+	}
+
+	rows, err := db.sql.Queryx(`
+SELECT 
+  jsonb_object_keys(s.metrics->'avg_metrics') AS name,
+  max(s.end_time) AS end_time
+FROM trials t
+INNER JOIN steps s ON t.id=s.trial_id
+WHERE t.experiment_id=$1
+  AND s.end_time > $2
+GROUP BY name;`, experimentID, sStartTime)
+	if err != nil {
+		return nil, nil, sEndTime, vEndTime, errors.Wrapf(err,
+			"error querying training metric names for experiment %d", experimentID)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var row namesWrapper
+		err = rows.StructScan(&row)
+		if err != nil {
+			return nil, nil, sEndTime, vEndTime, errors.Wrapf(err,
+				"error scanning training metric names for experiment %d",
+				experimentID)
+		}
+		training = append(training, row.Name)
+		if row.EndTime.After(sEndTime) {
+			sEndTime = row.EndTime
+		}
+	}
+
+	rows, err = db.sql.Queryx(`
+SELECT
+  jsonb_object_keys(v.metrics->'validation_metrics') AS name,
+  max(v.end_time) AS end_time
+FROM trials t
+INNER JOIN steps s ON t.id=s.trial_id
+LEFT OUTER JOIN validations v ON s.id=v.step_id AND s.trial_id=v.trial_id
+WHERE t.experiment_id=$1
+  AND v.end_time > $2
+GROUP BY name;`, experimentID, vStartTime)
+	if err != nil {
+		return nil, nil, sEndTime, vEndTime, errors.Wrapf(err,
+			"error querying validation metric names for experiment %d", experimentID)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var row namesWrapper
+		err = rows.StructScan(&row)
+		if err != nil {
+			return nil, nil, sEndTime, vEndTime, errors.Wrapf(err,
+				"error scanning validation metric names for experiment %d",
+				experimentID)
+		}
+		validation = append(validation, row.Name)
+		if row.EndTime.After(sEndTime) {
+			sEndTime = row.EndTime
+		}
+	}
+
+	return training, validation, sEndTime, vEndTime, err
+}
+
+// MetricBatches returns the milestones (in batches processed) at which a specific metric was
+// recorded.
+func (db *PgDB) MetricBatches(experimentID int, trainingMetric string, validationMetric string,
+	startTime time.Time) (batches []int32, endTime time.Time, err error) {
+	endTime = startTime
+	var metricName string
```

Comment:
This validation is duplicating what's done in api_experiment.py. It's also common to some other API endpoints I plan on adding so I've refactored it out (in a subsequent PR) so it's only done once per API call and with shared code between all calls.

Reply:
Done.